chain/bitcoind.go

```diff
@@ -0,0 +1,1122 @@
+		}
+	}
+	return nil
+}
+
+// RescanBlocks rescans any blocks passed, returning only the blocks that
+// matched as []btcjson.BlockDetails.
+func (c *BitcoindClient) RescanBlocks(blockHashes []chainhash.Hash) (
+	[]btcjson.RescannedBlock, error) {
+
+	rescannedBlocks := make([]btcjson.RescannedBlock, 0, len(blockHashes))
+	for _, hash := range blockHashes {
+		header, err := c.GetBlockHeaderVerbose(&hash)
+		if err != nil {
+			log.Warnf("Unable to get header %s from bitcoind: %s",
+				hash, err)
+			continue
+		}
+
+		block, err := c.GetBlock(&hash)
+		if err != nil {
+			log.Warnf("Unable to get block %s from bitcoind: %s",
+				hash, err)
+			continue
+		}
+
+		relevantTxes, err := c.filterBlock(block, header.Height,
+			false)
+		if len(relevantTxes) > 0 {
+			rescannedBlock := btcjson.RescannedBlock{
+				Hash: hash.String(),
+			}
+			for _, tx := range relevantTxes {
+				rescannedBlock.Transactions = append(
+					rescannedBlock.Transactions,
+					hex.EncodeToString(tx.SerializedTx),
+				)
+			}
+			rescannedBlocks = append(rescannedBlocks,
+				rescannedBlock)
+		}
+	}
+	return rescannedBlocks, nil
+}
+
+// Rescan rescans from the block with the given hash until the current block,
+// after adding the passed addresses and outpoints to the client's watch list.
+func (c *BitcoindClient) Rescan(hash *chainhash.Hash, addrs []btcutil.Address,
+	outPoints []*wire.OutPoint) error {
+
+	// Update addresses.
+	select {
+	case c.rescanUpdate <- addrs:
+	case <-c.quit:
+		return nil
+	}
+
+	// Update outpoints.
+	select {
+	case c.rescanUpdate <- outPoints:
+	case <-c.quit:
+		return nil
+	}
+
+	// Kick off the rescan with a new hash.
+	select {
+	case c.rescanUpdate <- hash:
+	case <-c.quit:
+		return nil
+	}
+
+	return nil
+}
+
+// SendRawTransaction sends a raw transaction via bitcoind.
+func (c *BitcoindClient) SendRawTransaction(tx *wire.MsgTx,
+	allowHighFees bool) (*chainhash.Hash, error) {
+
+	return c.client.SendRawTransaction(tx, allowHighFees)
+}
+
+// Start attempts to establish a client connection with the remote server.
+// If successful, handler goroutines are started to process notifications
+// sent by the server.  After a limited number of connection attempts, this
+// function gives up, and therefore will not block forever waiting for the
+// connection to be established to a server that may not exist.
+func (c *BitcoindClient) Start() error {
+	// Verify that the server is running on the expected network.
+	net, err := c.GetCurrentNet()
+	if err != nil {
+		c.client.Disconnect()
+		return err
+	}
+	if net != c.chainParams.Net {
+		c.client.Disconnect()
+		return errors.New("mismatched networks")
+	}
+
+	c.quitMtx.Lock()
+	c.started = true
+	c.quitMtx.Unlock()
+
+	c.wg.Add(2)
+	go c.handler()
+	go c.socketHandler()
+	return nil
+}
+
+// Stop disconnects the client and signals the shutdown of all goroutines
+// started by Start.
+func (c *BitcoindClient) Stop() {
+	c.quitMtx.Lock()
+	select {
+	case <-c.quit:
```

Comment:
What's the point of this case? In what event would something be sent over the quit channel as the client is signaled to shutdown? 

Why not just use the typical pattern with atomic integers that we typically use within lnd?

Reply:
This is copied from the original `rpc.go` client for `btcd`. We can refactor this.